crates/bevy_ui/src/widget/button.rs

```diff
@@ -1,7 +1,58 @@
-use bevy_ecs::{prelude::Component, reflect::ReflectComponent};
+use bevy_ecs::prelude::*;
+use bevy_picking::events::{Down, Out, Over, Pointer, Up};
 use bevy_reflect::{std_traits::ReflectDefault, Reflect};
 
-/// Marker struct for buttons
+/// The data required to create a buttonlike widget that can respond to hovering and being clicked / pressed.
+///
+/// Information about whether or not this button is currently pressed or hovered is gathered from
+/// user input, processed via [`bevy_picking`] and read in the [`determine_button_interaction`] system.
+///
+/// See [`FocusPolicy`](crate::focus::FocusPolicy) to configure whether or not the button should block interactions with lower nodes.
 #[derive(Component, Debug, Default, Clone, Copy, PartialEq, Eq, Reflect)]
 #[reflect(Component, Default, Debug, PartialEq)]
-pub struct Button;
+pub struct Button {
+    /// Whether the button is currently pressed.
+    ///
+    /// This is `true` as long as the button is pressed, but only while the pointer
+    /// is over the button.
+    pub pressed: bool,
+    /// Whether the button is currently hovered.
+    ///
+    /// This is only `true` while the pointer is over this button,
+    /// and will not be set if there is no pointer.
+    pub hovered: bool,
+}
+
+/// An observer that watches for [`Pointer<Over>`] events and sets the [`Button::hovered`] field to `true`.
+pub fn button_hover_observer(
+    trigger: Trigger<Pointer<Over>>,
+    mut button_query: Query<&mut Button>,
+) {
+    if let Ok(mut button) = button_query.get_mut(trigger.entity()) {
+        button.hovered = true;
+    }
+}
+
+/// An observer that watches for [`Pointer<Out>`] events and sets the [`Button::hovered`] field to `false`.
+///
+/// [`Button::pressed`] is also set to `false`, to ensure that the button.
+pub fn button_out_observer(trigger: Trigger<Pointer<Out>>, mut button_query: Query<&mut Button>) {
+    if let Ok(mut button) = button_query.get_mut(trigger.entity()) {
+        button.hovered = false;
+        button.pressed = false;
```

Comment:
```suggestion
```
Apps can check for `hovered && pressed` but in web if you press and drag away off the button you will still see the "pressed down" effect.
It's also simpler, so easier to work with

Reply:
But this will require `Pointer<Down>` to affect this even if it's unrelated to this entity.

Reply:
Yeah, currently I'm encouraging a "handle button events as soon as buttons are pressed" design, which dodges this complexity IMO.

Reply:
Sure, if we go with Nth's idea, people can pretty easily reimplement this behavior :)